File: app/src/main/java/com/mdeiml/richard/CalculateEloTask.java
package com.mdeiml.richard;

import android.database.*;
import android.database.sqlite.*;
import android.content.Context;
import android.os.AsyncTask;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import org.apache.commons.math3.linear.OpenMapRealMatrix;
import org.apache.commons.math3.linear.ArrayRealVector;
import org.apache.commons.math3.linear.RealVector;
import org.apache.commons.math3.linear.SingularValueDecomposition;

public class CalculateEloTask extends AsyncTask<String, Void, HashMap<String, Double>> {

    private SavedMatchesDbHelper dbHelper;
    private String sqlPlayerStats;
    private WeakReference<NewGameActivity> activity;

    public CalculateEloTask(NewGameActivity activity) {
        this.dbHelper = new SavedMatchesDbHelper(activity);
        this.sqlPlayerStats = activity.getResources().getString(R.string.sql_player_stats);
        this.activity = new WeakReference(activity);
    }
    
    protected HashMap<String, Double> doInBackground(String... players) {
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        LinkedList<String> queue = new LinkedList<>();
        ArrayList<String> connectedPlayers = new ArrayList<>();
        HashMap<Pair<String, String>, Pair<Integer, Double>> connections = new HashMap<>();

        queue.addAll(Arrays.asList(players));
        while (!queue.isEmpty()) {
            String player = queue.remove(0);
            Cursor cursor = db.rawQuery(sqlPlayerStats, new String[] { player, player });
            int otherI = cursor.getColumnIndexOrThrow("player");
            int countI = cursor.getColumnIndexOrThrow("count");
            int winsI = cursor.getColumnIndexOrThrow("wins");
            while (cursor.moveToNext()) {
                String other = cursor.getString(otherI);
                if (connectedPlayers.contains(other)) {
                    continue;
                }
                int count = cursor.getInt(countI);
                if (count == 0) {
                    continue;
                }
                int wins = cursor.getInt(winsI);
                double prob = (wins + 0.5) / (count + 1); // Add one draw to account for to little data
                double res = 400 * Math.log10(1 / prob - 1);
                connections.put(new Pair<>(player, other), new Pair(count, res));

                queue.add(other);
            }
            cursor.close();
            connectedPlayers.add(player);
        }

        int rows = connectedPlayers.size() * (connectedPlayers.size() + 1) / 2;
        OpenMapRealMatrix matrix = new OpenMapRealMatrix(rows, connectedPlayers.size());
        ArrayRealVector b = new ArrayRealVector(rows);

        int row = 0;
        for (int i = 0; i < connectedPlayers.size(); i++) {
            for (int j = i + 1; j < connectedPlayers.size(); j++) {
                String name1 = connectedPlayers.get(i);
                String name2 = connectedPlayers.get(j);
                Pair<Integer, Double> connection = connections.get(new Pair(name1, name2));
                if (connection == null) {
                    connection = connections.get(new Pair(name2, name1));
                    if (connection == null) {
                        connection = new Pair(1, 0.0); // Add one draw
                    }
                }
                int w = connection.getFirst();
                double r = connection.getSecond();
                matrix.setEntry(row, i, -w);
                matrix.setEntry(row, j, w);
                b.setEntry(row, w * r);
                row++;
            }
        }
        
        SingularValueDecomposition svd = new SingularValueDecomposition(matrix);
        RealVector x = svd.getSolver().solve(b);

        HashMap<String, Double> result = new HashMap<>();
        for (int i = 0; i < connectedPlayers.size(); i++) {
            result.put(connectedPlayers.get(i), x.getEntry(i));
        }
        
        return result;
    }

    public void onPostExecute(HashMap<String, Double> result) {
        NewGameActivity a = activity.get();
        if (a != null) {
            a.addElos(result);
        }
    }
}

File: app/src/main/kotlin/com/mdeiml/richard/CalculateEloTask.kt
package com.mdeiml.richard

import android.database.*
import android.database.sqlite.*
import android.content.Context
import android.os.AsyncTask
import java.lang.ref.WeakReference
import java.util.LinkedList
import org.apache.commons.math3.linear.OpenMapRealMatrix
import org.apache.commons.math3.linear.ArrayRealVector
import org.apache.commons.math3.linear.RealVector
import org.apache.commons.math3.linear.SingularValueDecomposition

class CalculateEloTask(activity: NewGameActivity) : AsyncTask<String, Void, HashMap<String, Double>>() {

    private val dbHelper = SavedMatchesDbHelper(activity)
    private val sqlPlayerStats = activity.getResources().getString(R.string.sql_player_stats)
    private val ref = WeakReference(activity)
    
    override fun doInBackground(vararg players: String): HashMap<String, Double> {
        val db = dbHelper.getReadableDatabase()

        val queue = LinkedList<String>()
        val connectedPlayers = ArrayList<String>()
        val connections = HashMap<Pair<String, String>, Pair<Int, Double>>()

        queue.addAll(players)
        while (!queue.isEmpty()) {
            val player = queue.removeAt(0)
            val cursor = db.rawQuery(sqlPlayerStats, arrayOf(player, player))
            val otherI = cursor.getColumnIndexOrThrow("player")
            val countI = cursor.getColumnIndexOrThrow("count")
            val winsI = cursor.getColumnIndexOrThrow("wins")
            while (cursor.moveToNext()) {
                val other = cursor.getString(otherI)
                if (connectedPlayers.contains(other)) {
                    continue
                }
                val count = cursor.getInt(countI)
                if (count == 0) {
                    continue
                }
                val wins = cursor.getInt(winsI)
                val prob = (wins + 0.5) / (count + 1.0) // Add one draw to account for to little data
                val res = 400.0 * Math.log10(1.0 / prob - 1.0)
                connections.put(Pair(player, other), Pair(count, res))

                queue.add(other)
            }
            cursor.close()
            connectedPlayers.add(player)
        }

        val rows = connectedPlayers.size * (connectedPlayers.size + 1) / 2
        val matrix = OpenMapRealMatrix(rows, connectedPlayers.size)
        val b = ArrayRealVector(rows)

        var row = 0
        for ((i, name1) in connectedPlayers.withIndex()) {
            for ((j, name2) in connectedPlayers.drop(i).withIndex()) {
                val connection = connections.get(Pair(name1, name2)) ?: connections.get(Pair(name2, name1)) ?: Pair(1, 0.0)
                val (w, r) = connection
                matrix.setEntry(row, i, -w.toDouble())
                matrix.setEntry(row, j, w.toDouble())
                b.setEntry(row, w * r)
                row++
            }
        }
        
        val svd = SingularValueDecomposition(matrix)
        val x = svd.getSolver().solve(b)

        val result = HashMap<String, Double>()
        for ((i, name) in connectedPlayers.withIndex()) {
            result.put(name, x.getEntry(i))
        }
        
        return result
    }

    override fun onPostExecute(result: HashMap<String, Double>) {
        val a = ref.get()
        if (a != null) {
            a.addElos(result)
        }
    }
}
